lib/determinator/tracking.rb

```diff
@@ -0,0 +1,62 @@
+require 'determinator/tracking/tracker'
+require 'determinator/tracking/context'
+
+module Determinator
+  module Tracking
+    class << self
+      def instance
+        Thread.current[:determinator_tracker]
+      end
+
+      def start!(type)
+        Thread.current[:determinator_tracker] = Tracker.new(type)
```

Comment:
Afaik there's no guarantee that every part of a rails request will execute in the same thread.

This gem tries to address it by using a singleton, perhaps bring it in as a dependency or use as inspiration.

Reply:
Forgot the link 😱 https://github.com/ElMassimo/request_store_rails

Reply:
That's a very good point! However that gem requires each spawned thread to remember to call `RequestLocals.set_current_store_id`, which is probably unreasonable to expect of all users of this gem. I think doing determinations in a spawned thread is enough of an edge case that we can live with it? I'll make sure to point that out in the readme though.

Reply:
@danielcooper what's an example scenario where a rails (or rack) request is handled by two different threads? (curious and I don't know anything about Puma's multi-threading model)

Reply:
I don't believe puma would affect this as it spins up new threads as requests come in, presumably before the rack chain is started.

[This](https://guides.rubyonrails.org/threading_and_code_execution.html) suggests that each request be will provided a controller instance. That's probably the best thing to hang this determination recorder off of? 

My concern isn't so much the rails stack is doing, but what application code is doing. It wouldn't be that remarkable to spin up a few threads to perform some IO operation (fan out?).

Having said all that, it's probably fine for most cases - and when it's not the cost of getting it wrong is just some strange stats.

Reply:
> This suggests that each request be will provided a controller instance. That's probably the best thing to hang this determination recorder off of?

We don't really have the controller instance available here. We could maybe pass that (or the request object) on determinations, but that feels way more invasive.

> My concern isn't so much the rails stack is doing, but what application code is doing. It wouldn't be that remarkable to spin up a few threads to perform some IO operation (fan out?).

Yeah that is fair... Hopefully doing determinations in those threads is not too likely. But I'll add a note to point that out.